File: apps/kira-studio/internal/gitsearch/conformance_test.go
package gitsearch

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"reflect"
	"runtime"
	"testing"
)

// conformanceRow mirrors conformance.test.ts's own ConformanceRow shape exactly — one JSON file,
// two readers (docs/v1.3/plans/G23-search.md D9).
type conformanceRow struct {
	Name           string          `json:"name"`
	Query          conformanceQry  `json:"query"`
	Subject        string          `json:"subject"`
	Body           string          `json:"body"`
	AuthorName     string          `json:"authorName"`
	AuthorEmail    string          `json:"authorEmail"`
	CommitterName  string          `json:"committerName"`
	CommitterEmail string          `json:"committerEmail"`
	SHA            string          `json:"sha"`
	Expect         conformanceWant `json:"expect"`
}

type conformanceQry struct {
	Text          string `json:"text"`
	CaseSensitive bool   `json:"caseSensitive"`
	WholeWord     bool   `json:"wholeWord"`
	Regex         bool   `json:"regex"`
}

type conformanceWant struct {
	Supported bool     `json:"supported"`
	Fields    []string `json:"fields"`
}

type conformanceCorpus struct {
	Rows []conformanceRow `json:"rows"`
}

// corpusPath resolves packages/git-core/testdata/searchConformance.json relative to this source
// file's own location — stable regardless of the working directory `go test` is invoked from.
func corpusPath(t *testing.T) string {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed")
	}
	// internal/gitsearch -> apps/kira-studio -> apps -> <repo root> -> packages/git-core/testdata
	return filepath.Join(thisFile, "..", "..", "..", "..", "..", "packages", "git-core", "testdata", "searchConformance.json")
}

func loadCorpus(t *testing.T) conformanceCorpus {
	t.Helper()
	b, err := os.ReadFile(corpusPath(t))
	if err != nil {
		t.Fatalf("read searchConformance.json: %v", err)
	}
	var corpus conformanceCorpus
	if err := json.Unmarshal(b, &corpus); err != nil {
		t.Fatalf("unmarshal searchConformance.json: %v", err)
	}
	return corpus
}

func toQuery(q conformanceQry) Query {
	return Query{Text: q.Text, CaseSensitive: q.CaseSensitive, WholeWord: q.WholeWord, Regex: q.Regex}
}

func toFields(row conformanceRow) CommitFields {
	return CommitFields{
		SHA: row.SHA, Subject: row.Subject, Body: row.Body,
		AuthorName: row.AuthorName, AuthorEmail: row.AuthorEmail,
		CommitterName: row.CommitterName, CommitterEmail: row.CommitterEmail,
	}
}

// TestConformanceCorpus_NonEmpty guards against a silently-empty or unreadable corpus file
// passing this suite vacuously.
func TestConformanceCorpus_NonEmpty(t *testing.T) {
	corpus := loadCorpus(t)
	if len(corpus.Rows) == 0 {
		t.Fatal("searchConformance.json has no rows")
	}
}

// TestConformanceCorpus_AgreesWithCompileAndMatchFields is G23's own exit-criterion item 2: every
// row in the shared corpus, run through Compile + MatchFields, agrees with the expectation
// conformance.test.ts asserts against the exact same file on the TypeScript side.
func TestConformanceCorpus_AgreesWithCompileAndMatchFields(t *testing.T) {
	corpus := loadCorpus(t)
	for _, row := range corpus.Rows {
		row := row
		t.Run(row.Name, func(t *testing.T) {
			m, err := Compile(toQuery(row.Query))
			if !row.Expect.Supported {
				if !errors.Is(err, ErrUnsupportedPattern) {
					t.Fatalf("Compile error = %v, want ErrUnsupportedPattern", err)
				}
				return
			}
			if err != nil {
				t.Fatalf("Compile: unexpected error %v", err)
			}
			got := m.MatchFields(toFields(row))
			gotStrings := fieldStrings(got)
			want := row.Expect.Fields
			if want == nil {
				want = []string{}
			}
			if !reflect.DeepEqual(gotStrings, want) {
				t.Fatalf("MatchFields = %v, want %v", gotStrings, want)
			}
		})
	}
}

File: packages/git-core/src/search/conformance.test.ts
import { describe, expect, test } from 'bun:test';
import path from 'node:path';
import type { CommitFields, SearchField } from './matcher.ts';
import { matchCommitFields } from './matcher.ts';
import type { SearchQuery } from './query.ts';
import { compileQuery } from './query.ts';

/**
 * `docs/plans/G23-search.md` D9: the shared conformance corpus, read from BOTH languages —
 * `apps/kira-studio/internal/gitsearch/conformance_test.go` reads the exact same JSON and runs
 * `Compile`/`MatchFields`. A row failing here and passing there (or vice versa) is exactly the
 * silent divergence SPEC's own open item (`docs/v1.3/SPEC.md:446-450`) forbids.
 */

interface ConformanceQuery {
  readonly text: string;
  readonly caseSensitive: boolean;
  readonly wholeWord: boolean;
  readonly regex: boolean;
}

interface ConformanceRow {
  readonly name: string;
  readonly query: ConformanceQuery;
  readonly subject: string;
  readonly body: string;
  readonly authorName: string;
  readonly authorEmail: string;
  readonly committerName: string;
  readonly committerEmail: string;
  readonly sha: string;
  readonly expect: {
    readonly supported: boolean;
    readonly fields: readonly SearchField[];
  };
}

interface ConformanceCorpus {
  readonly rows: readonly ConformanceRow[];
}

const fixturePath = path.join(import.meta.dir, '..', '..', 'testdata', 'searchConformance.json');

async function loadCorpus(): Promise<ConformanceCorpus> {
  const text = await Bun.file(fixturePath).text();
  return JSON.parse(text) as ConformanceCorpus;
}

function toSearchQuery(q: ConformanceQuery): SearchQuery {
  return { ...q, scope: 'commits' };
}

function toCommitFields(row: ConformanceRow): CommitFields {
  return {
    sha: row.sha,
    subject: row.subject,
    body: row.body,
    authorName: row.authorName,
    authorEmail: row.authorEmail,
    committerName: row.committerName,
    committerEmail: row.committerEmail,
  };
}

describe('search conformance corpus (git-core side)', () => {
  test('the corpus file parses and is non-empty', async () => {
    const corpus = await loadCorpus();
    expect(corpus.rows.length).toBeGreaterThan(0);
  });

  test('every row agrees with compileQuery + matchCommitFields', async () => {
    const corpus = await loadCorpus();
    for (const row of corpus.rows) {
      const compiled = compileQuery(toSearchQuery(row.query));
      if (!row.expect.supported) {
        // JS has no RE2 to reject a pattern against -- an "unsupported in Go" row must still
        // compile fine as JS (that is the whole premise of D6's honest unsupportedPattern
        // member: the pattern IS valid JavaScript). This corpus does not assert anything about
        // what compileQuery itself returns for such a row beyond "it must not be 'invalid'".
        expect(compiled.kind, `row ${row.name}: JS must be able to compile this pattern`).toBe(
          'ok',
        );
        continue;
      }
      expect(compiled.kind, `row ${row.name}: expected an 'ok' compiled query`).toBe('ok');
      if (compiled.kind !== 'ok') continue;
      const got = matchCommitFields(toCommitFields(row), compiled);
      expect(got, `row ${row.name}`).toEqual(row.expect.fields);
    }
  });
});
